cmd/snap-confine/snap-confine.c

```diff
@@ -301,18 +302,25 @@ int main(int argc, char **argv)
 				}
 			}
 
-			// Associate each snap process with a dedicated snap freezer
-			// control group. This simplifies testing if any processes
-			// belonging to a given snap are still alive.
-			// See the documentation of the function for details.
+			// Associate each snap process with a dedicated snap freezer cgroup
+			// and snap pids cgroup. All snap processes belonging to one snap
+			// share the freezer cgroup. All snap processes belonging to one
+			// app or one hook share the pids cgroup.
+			//
+			// This simplifies testing if any processes belonging to a given
+			// snap are still alive as well as to properly account for each
+			// application and service.
 			if (getegid() != 0 && saved_gid == 0) {
-				// Temporarily raise egid so we can chown the freezer cgroup
-				// under LXD.
+				// Temporarily raise egid so we can chown the pids and freezer
+				// cgroups under LXD.
 				if (setegid(0) != 0) {
 					die("cannot set effective group id to root");
 				}
 			}
 			sc_cgroup_freezer_join(snap_instance, getpid());
+			if (sc_feature_enabled(SC_FEATURE_REFRESH_APP_AWARENESS)) {
+				sc_cgroup_pids_join(security_tag, getpid());
```

Comment:
Why are you differentiating between each command and hook? Isn't it sufficient to add them all to the same pids cgroup? Does this have to do with the algorithm for the soft/hard checks that haven't landed yet? If so, perhaps expand the comment a little to make this clear (not the details of the algorithm, just something like 'Because of the way we track pids in our refresh checks, we break out each command and hook into its own pids cgroup'.

That said, since snaps can have tens of commands and there can be scores of snaps on the system, does it make sense resource-wise to break the pids cgroups up like this?

Reply:
We want to have accurate information about apps and hooks, the main motivation is to perform the refresh checks but I think this will become more important as we move to cgroups v2 with their enforced hierarchy model. To impose command level resource restrictions we will need command-level granularity.

We could bundle apps, hooks and enduring services into bucket A and non-enduring services to bucket B but I think this is premature optimisation. Systemd uses loads of cgroup nodes for myriad of services, slices, users and what not. There is no indication that this is resource intensive in itself.

Another aspect is that we will be able to remove empty cgroups. They will exist only briefly after termination.

Lastly this is all behind a feature flag: we are free to change the design if something better can be done.